examples/se_2_3_localization.cpp

```diff
@@ -0,0 +1,302 @@
+/**
+ * \file se_2_3_localization.cpp
+ *
+ *  Created on: Aug 03, 2020
+ *     \author: prashanthr05
+ *
+ *  ---------------------------------------------------------
+ *  This file is:
+ *  (c) 2018 Joan Sola @ IRI-CSIC, Barcelona, Catalonia
+ *
+ *  This file is part of `manif`, a C++ template-only library
+ *  for Lie theory targeted at estimation for robotics.
+ *  Manif is:
+ *  (c) 2018 Jeremie Deray @ IRI-UPC, Barcelona
+ *  ---------------------------------------------------------
+ *
+ *  ---------------------------------------------------------
+ *  Demonstration example:
+ *
+ *  3D Robot localization and linear velocity estimation based on fixed beacons.
+ *
+ *  ---------------------------------------------------------
+ *
+ *  We consider a robot in 3D space surrounded by a small
+ *  number of punctual landmarks or _beacons_.
+ *  The robot receives control actions in the form of axial
+ *  and angular velocities, and is able to measure the location
+ *  of the beacons w.r.t its own reference frame.
+ *
+ *  The robot extended pose X is in SE_2(3) and the beacon positions b_k in R^3,
+ *
+ *      X = |  R   t  v|              // position, orientation and linear velocity
+ *          |  0   1  0|
+ *          |  0   0  1|
+ *
+ *      b_k = (bx_k, by_k, bz_k)    // lmk coordinates in world frame
+ *
+ * Note that the linear velocity is expressed in a frame whose origin coincides with
+ * the robot frame but has a orientation which is same as the world frame
+ *
+ *  The control signal u is a vector in se_2(3) comprising longitudinal
+ *  velocity vx and angular velocity wz, with no other velocity
+ *  components or acceleration components, integrated over the sampling time dt.
+ *
+ *      u = (vx*dt, 0, 0, 0, 0, w*dt, 0, 0, 0)
+ *
```

Comment:
I guess control `u` should be IMU measurements

Reply:
I was actually wondering about it; whether to pass the IMU measurements as exogenous inputs or not.
I rather chose this simpler case instead. 

I will modify this to use expected IMU measurements.

Reply:
I do not know the mechanisms of this SE23 group. We had invented a different Lie group for IMU, see here https://upcommons.upc.edu/handle/2117/187253, where the time interval is part of the group too.

In our case, integrating motion is done as follows:

1. `u = [w*dt, 0, a*dt, dt] \in R^10`
2. `X(t+dt) = X(t) * Exp(u)`

In your case I do not know how you handle time so I cannot suggest an implementation for this part of the code commented here. 

Nevertheless the idea of these examples is that they be realistic, and I think that the inputs to this example should be the IMU measurements `a` and `w`, together with the time interval `dt`.

Reply:
Improved the example in https://github.com/artivis/manif/pull/154/commits/945c4728639519d7eaa7bb47e7de7eacb9171fab.

using discrete-time propagation compatible with SE_2(3) composition,
```
p_{k+1} = p_{k} + R_{k} (R_{k}.T v_{k} dt + 0.5*dt*dt*acc_{k})
R_{k+1} = R_{k} Exp_SO3(omega_{k} dt + w_omega)
v_{k+1} = v_{k} + R_{k} (acc_{k} dt + w_acc)
```

The input vector `u` becomes state-dependent. 
The measurements remain the same beacon measurements.

Please let me know what you think about it.

For IMU measurements, 
```
alpha << 10.0, 0.0, 9.80665;
omega << 1, 0.1, 0;
g << 0, 0, -9.80665;
```

The results are as follows, 

``` sh
X STATE     :    X      Y      Z    TH_x   TH_y   TH_z     V_x   V_y    V_z
---------------------------------------------------------------------------
X initial   : +0.000 +0.000 +0.000 +0.000 +0.000 +0.000 +0.000 +0.000 +0.000
---------------------------------------------------------------------------
X simulated : +0.000 +0.000 +0.000 +0.000 +0.000 +0.000 +0.000 +0.000 +0.000
X estimated : +0.004 +0.009 -0.002 +0.005 -0.000 -0.005 -0.033 +0.054 -0.044
X unfilterd : +0.000 +0.000 +0.000 +0.060 +0.082 -0.060 -0.033 +0.054 -0.044
---------------------------------------------------------------------------
X simulated : +0.050 +0.000 +0.000 +0.100 +0.010 +0.000 +1.000 +0.000 +0.000
X estimated : +0.054 +0.008 -0.005 +0.099 +0.009 -0.005 +0.969 +0.116 -0.023
X unfilterd : +0.050 -0.001 -0.002 +0.082 +0.193 -0.114 +0.974 +0.090 -0.064
---------------------------------------------------------------------------
X simulated : +0.101 -0.005 -0.000 +0.200 +0.020 +0.000 +2.010 -0.098 +0.000
X estimated : +0.104 -0.001 -0.002 +0.201 +0.018 -0.004 +2.056 -0.010 -0.120
X unfilterd : +0.100 -0.010 -0.008 +0.100 +0.139 -0.086 +2.039 -0.112 -0.282
---------------------------------------------------------------------------
X simulated : +0.152 -0.015 +0.000 +0.300 +0.030 +0.000 +3.029 -0.295 +0.005
X estimated : +0.153 -0.012 -0.004 +0.299 +0.029 -0.000 +3.104 -0.250 -0.081
X unfilterd : +0.150 -0.025 -0.012 +0.245 +0.097 -0.050 +3.067 -0.452 -0.336
---------------------------------------------------------------------------
X simulated : +0.203 -0.029 +0.001 +0.400 +0.040 -0.000 +4.059 -0.590 +0.020
X estimated : +0.204 -0.026 -0.001 +0.401 +0.040 -0.001 +4.222 -0.454 -0.129
X unfilterd : +0.202 -0.038 -0.012 +0.425 +0.106 -0.114 +4.193 -0.622 -0.390
---------------------------------------------------------------------------
X simulated : +0.255 -0.049 +0.002 +0.500 +0.050 -0.000 +5.098 -0.984 +0.050
X estimated : +0.256 -0.046 +0.001 +0.502 +0.049 -0.001 +5.299 -0.889 -0.065
X unfilterd : +0.251 -0.071 -0.002 +0.587 +0.185 -0.014 +5.233 -1.346 -0.160
---------------------------------------------------------------------------
X simulated : +0.308 -0.074 +0.005 +0.600 +0.060 -0.000 +6.148 -1.477 +0.100
X estimated : +0.310 -0.070 +0.002 +0.598 +0.059 -0.002 +6.373 -1.314 -0.085
X unfilterd : +0.300 -0.109 +0.001 +0.742 +0.198 +0.076 +6.243 -2.041 -0.174
---------------------------------------------------------------------------
X simulated : +0.361 -0.104 +0.009 +0.700 +0.070 -0.000 +7.207 -2.070 +0.175
X estimated : +0.362 -0.102 +0.007 +0.701 +0.070 -0.003 +7.451 -1.946 -0.099
X unfilterd : +0.350 -0.145 -0.022 +0.766 +0.082 +0.121 +7.235 -2.821 -0.724
---------------------------------------------------------------------------
X simulated : +0.414 -0.138 +0.014 +0.800 +0.080 -0.000 +8.276 -2.762 +0.281
X estimated : +0.415 -0.137 +0.011 +0.800 +0.078 -0.002 +8.597 -2.625 +0.110
X unfilterd : +0.399 -0.197 -0.003 +0.915 +0.132 +0.228 +8.317 -3.862 -0.234
---------------------------------------------------------------------------
X simulated : +0.468 -0.178 +0.021 +0.900 +0.090 -0.000 +9.355 -3.555 +0.424
X estimated : +0.469 -0.178 +0.019 +0.900 +0.090 +0.000 +9.675 -3.328 +0.249
X unfilterd : +0.456 -0.236 -0.005 +1.048 +0.092 +0.228 +9.425 -4.529 -0.293
---------------------------------------------------------------------------

```

Reply:
> We had invented a different Lie group for IMU, see here https://hal.archives-ouvertes.fr/hal-02183498/document, where the time interval is part of the group too.

Really interesting work.

Reply:
@artivis @joansola would you have any feedback on this updated example?

Reply:
Let me have a look --